File: tiny-interpreter/src/parser.ts
import * as lexer from "./lexer";
import { ASTNode, ASTNodeType, Token, TokenType as TType, TokenStream } from "./models";

export function parse(code: string) {
  const tokens = lexer.tokenize(code);
  const rootNode = prog(tokens);
  return rootNode;
}

/**
 * AST根节点
 * @param tokens
 */
function prog(tokens: Token[]): ASTNode {
  const ts = new TokenStream(tokens);
  let node = new ASTNode(ASTNodeType.Program, null);
  let token;
  while (token = ts.peekToken()) {
    if (token.type == TType.EOF) break;
    let child = statement(ts);
    if (!child) continue;
    node.addChild(child)
  }
  return node;
}

function statement(ts: TokenStream) {
  let child = varDeclare(ts)
    || assignmentStatement(ts)
    || expressionStatement(ts)
    || blockStatement(ts)
    || returnStatement(ts)
    || functionDefinition(ts)
    || callExpression(ts);
  return statementEnd(child, ts);
}

/**
 *
 * 函数声明语句 functionDefinition :: function Identifier '(' Identifier (',' Identifier)* ')' blockStatement
 */
function functionDefinition(ts: TokenStream): ASTNode | null {
  let token = ts.getToken();
  if (notToken(token, TType.FUNCTION)) {
    ts.unGetToken(token);
    return null;
  }
  token = ts.getToken();
  if (token.type !== TType.NAME) throw Error("function must has name");
  let node = new ASTNode(ASTNodeType.Function, token);
  token = ts.getToken();
  if (token.type !== TType.LP) throw Error("function define error");
  //函数定义参数不做签名检查，不用保留参数
  while (token = ts.getToken()) {
    if (token.type === TType.RP) break;
    if (token.type != TType.COMMA) throw Error("function args define error");
    token = ts.getToken();
    if (token.type !== TType.NAME) throw Error("function args define error");
  }
  token = ts.peekToken();
  if (token?.type != TType.LC) throw Error("function missing body")
  let child = blockStatement(ts);
  node.addChild(child);
  return node;
}

/**
 * 函数调用声明 callExpression :: Identifier '(' primary (',',primary) ')'
 */
function callExpression(ts: TokenStream) {
  let token = ts.getToken();
  if (notToken(token, TType.NAME)) {
    ts.unGetToken(token);
    return null;
  }
  if (notToken(ts.peekToken(), TType.LP)) {
    ts.unGetToken(token);
    return null;
  }
  let node = new ASTNode(ASTNodeType.Call, token);
  ts.getToken();
  token = ts.peekToken();
  if (!notToken(token, TType.RP)) {
    ts.getToken();
    return node;
  }
  let child = primary(ts);
  if (child) node.addChild(child);
  while (notToken(ts.peekToken(), TType.RP)) {
    if (ts.getToken()?.type != TType.COMMA) throw Error("call args syntax error");
    child = primary(ts);
    node.addChild(child);
  }
  ts.getToken();
  return node;
}

/**
 * 块级语句 blackStatement :: '{' (statement)* '}'
 */
function blockStatement(ts: TokenStream): ASTNode | null {
  let token = ts.getToken();
  if (notToken(token, TType.LC)) {
    ts.unGetToken(token);
    return null;
  }
  let node = new ASTNode(ASTNodeType.Block, null);
  do {
    let child = statement(ts);
    if (!child) break;
    node.addChild(child);
    token = ts.peekToken();
  } while (token && token.type !== TType.RC)
  ts.getToken();
  return node;
}

/**
 * 表达式语句 returnStatement :: return expressionStatement?
 */
function returnStatement(ts: TokenStream): ASTNode | null {
  let token = ts.getToken();
  if (notToken(token, TType.RETURN)) {
    ts.unGetToken(token);
    return null;
  }
  let node = new ASTNode(ASTNodeType.Return, null);
  const child = expressionStatement(ts);
  if (child) node.addChild(child);
  return node;
}

/**
 * 表达式语句 expressionStatement :: additiveExpression | callExpression
 */
function expressionStatement(ts: TokenStream): ASTNode | null {
  return callExpression(ts) || unaryExpression(ts) || binaryExpression(ts);
}

/**
 * 一元前缀表达式 unaryExpression :: [!,~] primary
 */
function unaryExpression(ts: TokenStream) {
  const token = ts.getToken();

  if (notToken(token, TType.UNARYOP)) {
    ts.unGetToken(token);
    return null;
  }
  let node = new ASTNode(ASTNodeType.Unary, token);
  let child = primary(ts);
  node.addChild(child);
  return node;
}

/**
 * 二元表达式 BinaryExpression :: [<<,>>,==,!=,<,<=,>,>=,additive]
 */
function binaryExpression(ts: TokenStream) {
  let child = additive(ts);
  if (!child) return null;
  if (child.type == ASTNodeType.Binary) return child;
  //证明当前node是number/string
  let token = ts.peekToken();
  switch (token?.type) {
    case TType.SHOP://<< >>
    case TType.RELOP://< > <= >=
    case TType.EQOP://== !=
    case TType.OR:// ||
    case TType.AND:// &&
    case TType.BITOR:// |
    case TType.BITXOR:// ^
    case TType.BITAND:// &
      {
        ts.getToken();
        let child2 = additive(ts);
        let node = new ASTNode(ASTNodeType.Binary, token);
        node.addChild(child);
        node.addChild(child2);
        return node;
      }
    default:
      return child;
  }
}

/**
 * 赋值语句
 * assignment :: Identifier '=' expressionStatement;
 * @param tokens
 */
function assignmentStatement(ts: TokenStream): ASTNode | null {
  let node: ASTNode | null = null;
  let token = ts.getToken();
  if (notToken(token, TType.NAME)) {
    ts.unGetToken(token);
    return null;
  }
  node = new ASTNode(ASTNodeType.AssignmentStmt, token);
  token = ts.getToken();
  if (notToken(token, TType.ASSIGN)) {
    ts.unGetToken(token);//+ =
    ts.unGetToken(token);//+ Identifier
    return null;
  }
  let child: ASTNode = expressionStatement(ts); // - additive
  if (!child) throw Error("invalidate assignment statement, expecting an expression");
  node.addChild(child);
  return node;
}

/**
 * var变量声明语句
 * varDeclaration :: 'var' Id ( '=' expressionStatement)?
 * @param tokens
 */
function varDeclare(ts: TokenStream): ASTNode | null {
  let node: ASTNode | null = null;
  let token = ts.getToken();
  if (notToken(token, TType.VAR)) {
    ts.unGetToken(token);
    return null;
  }
  token = ts.getToken();
  if (notToken(token, TType.NAME)) throw Error("variable name expected");
  node = new ASTNode(ASTNodeType.VarDeclaration, token);
  token = ts.getToken();
  if (notToken(token, TType.ASSIGN)) {
    ts.unGetToken(token);
    return node;
  }
  let child: ASTNode = expressionStatement(ts); // - additive
  if (!child) throw Error("invalide variable initialization, expecting an expression");
  node.addChild(child);
  return node;
}

/**
 * 加减法表达式：+-属于加
 * additiveExpression :: multiplicative ( '+'|'-' multiplicative)*
 * @param tokens
 */
function additive(ts: TokenStream): ASTNode | null {
  let child1 = multiplicative(ts);
  if (!child1) return null;

  let node = child1;
  let token = null;
  while (token = ts.getToken()) { //应用()*规则
    if (token.type != TType.PLUS && token.type != TType.MINUS) break;
    let child2 = multiplicative(ts);
    if (!child2) throw Error("invalid additive expression, expecting the right part.");
    node = new ASTNode(ASTNodeType.Binary, token);
    node.addChild(child1);
    node.addChild(child2)
    child1 = node;
  }
  ts.unGetToken(token);
  return node;
}

/**
 * 乘除法表达式
 * multiplicative :: primary ( '*'|'/' primary)*
 * @param tokens
 */
function multiplicative(ts: TokenStream): ASTNode | null {
  let child1 = primary(ts);
  let node = child1;
  if (!child1) return null;
  let token = null;
  while (token = ts.getToken()) { //应用()*规则
    if (notToken(token, TType.MULOP)) break;
    let child2 = primary(ts);
    if (!child2) throw Error("invalid additive expression, expecting the right part.");
    node = new ASTNode(ASTNodeType.Binary, token);
    node.addChild(child1);
    node.addChild(child2)
    child1 = node;
  }
  ts.unGetToken(token);
  return node;
}

/**
 * 基础表达式
 * primary :: NumberLiteral | Identifier | '(' expressionStatement ')'
 * @param tokens
 */
function primary(ts: TokenStream): ASTNode | null {
  let node: ASTNode | null = null;
  const token = ts.getToken();
  if (!token) return null;
  if (token.type == TType.PRIMARY) {
    node = new ASTNode(ASTNodeType.Primary, token);
  } else if (token.type == TType.NUMBER) {
    node = new ASTNode(ASTNodeType.NumberLiteral, token);
  } else if (token.type == TType.NAME) {
    node = new ASTNode(ASTNodeType.Identifier, token);
  } else if (token.type == TType.LP) {
    node = expressionStatement(ts);
    if (!node) throw Error("expecting an additive expression inside parenthesis");
    if (ts.getToken()?.type == TType.RP) {
    } else throw Error("expecting right parenthesis");
  } else {
    ts.unGetToken(token);
  }
  return node;
}

/**
 * 语句结尾 消化;
 * @param node
 * @param tokens
 */
function statementEnd(node: ASTNode, ts: TokenStream) {
  let token = ts.peekToken();
  if (token.type == TType.SEMI) {
    ts.getToken();
  }
  return node;
}

function notToken(token: Token | null, ...types: TType[]) {
  if (!token) return true;
  for (let type of types) {
    if (type == token.type) return false;
  }
  return true;
}

File: cpp-tiny-interpreter/parser.cpp
#include "models.h"
#include "vector"
#include "iostream"
#define ASSET_TS(tokens, tt)      \
  if (tokens.empty())             \
    return nullptr;               \
  if (tokens.front()->type != tt) \
    return nullptr;
#define IS_TYPE(tokens, tt) \
  (!tokens.empty() && tokens.front()->type == tt)
#define LOG_FRONT(tag, tokens) \
  cout << tag << " " << (tokens.empty() ? "" : tokens.front()->text) << endl;

vector<long> getRootFuncs();
ASTNode *expressionStatement(list<Token *> &tokens);
ASTNode *statement(list<Token *> &tokens);
typedef ASTNode *(*Statement)(list<Token *> &); //动态表达式函数类型定义
ASTNode *parser(list<Token *> &tokens)
{
  auto rootNode = new ASTNode(ASTNodeType::Program, "pwc");
  while (!tokens.empty())
  {
    auto child = statement(tokens);
    if (child != nullptr)
      rootNode->addChild(child);
  }
  return rootNode;
}
ASTNode *statement(list<Token *> &tokens)
{
  LOG_FRONT("statement", tokens);
  auto funclist = getRootFuncs();
  for (auto func : funclist)
  {
    auto child = ((Statement)func)(tokens);
    if (!tokens.empty() && tokens.front()->type == TokenType::SEMI)
      tokens.pop_front();
    if (child != nullptr)
      return child;
  };
  return nullptr;
}
/**
 * var变量声明语句
 * varDeclaration :: 'var' Id ( '=' expressionStatement)?
 * @param tokens
 */
ASTNode *varDeclare(list<Token *> &tokens)
{
  LOG_FRONT("varDeclare", tokens);
  ASSET_TS(tokens, TokenType::VAR);
  tokens.pop_front();
  if (!IS_TYPE(tokens, TokenType::NAME))
    throw "variable name expected";
  auto node = new ASTNode(ASTNodeType::VarDeclaration, tokens.front()->text);
  tokens.pop_front();
  if (!IS_TYPE(tokens, TokenType::ASSIGN))
    return node;
  tokens.pop_front();
  auto child = expressionStatement(tokens);
  if (child == nullptr)
    throw "invalide variable initialization, expecting an expression";
  node->addChild(child);
  return node;
}
/**
 * 赋值语句
 * assignment :: Identifier '=' expressionStatement;
 * @param tokens
 */
ASTNode *assignmentStatement(list<Token *> &tokens)
{
  LOG_FRONT("assignmentStatement", tokens);
  ASSET_TS(tokens, TokenType::NAME);
  auto nameToken = tokens.front();
  tokens.pop_front();
  if (!IS_TYPE(tokens, TokenType::ASSIGN))
  {
    tokens.push_front(nameToken);
    return nullptr;
  }
  tokens.pop_front();
  auto node = new ASTNode(ASTNodeType::AssignmentStmt, nameToken->text);
  auto child = expressionStatement(tokens);
  if (child != nullptr)
    node->addChild(child);
  return node;
};
/**
 * 基础表达式
 * primary :: NumberLiteral | Identifier | '(' expressionStatement ')'
 * @param tokens
 */
ASTNode *primary(list<Token *> &tokens)
{
  LOG_FRONT("primary", tokens);
  ASTNode *node = nullptr;
  if (IS_TYPE(tokens, TokenType::PRIMARY))
    node = new ASTNode(ASTNodeType::Primary, tokens.front()->text);
  if (IS_TYPE(tokens, TokenType::NUMBER))
    node = new ASTNode(ASTNodeType::NumberLiteral, tokens.front()->text);
  if (IS_TYPE(tokens, TokenType::NAME))
    node = new ASTNode(ASTNodeType::Identifier, tokens.front()->text);
  if (IS_TYPE(tokens, TokenType::LP))
  {
    tokens.pop_front();
    node = expressionStatement(tokens);
    if (node == nullptr)
      throw "expecting an additive expression inside parenthesis";
    if (tokens.front()->type != TokenType::RP)
      throw "expecting right parenthesis";
  }
  if (node != nullptr)
    tokens.pop_front();
  return node;
}
/**
 * 函数调用声明 callExpression :: Identifier '(' primary (',',primary) ')'
 */
ASTNode *callExpression(list<Token *> &tokens)
{
  LOG_FRONT("callExpression", tokens);
  ASSET_TS(tokens, TokenType::NAME);
  auto nameToken = tokens.front();
  tokens.pop_front();
  if (!IS_TYPE(tokens, TokenType::LP))
  {
    tokens.push_front(nameToken);
    return nullptr;
  }
  tokens.pop_front();
  auto node = new ASTNode(ASTNodeType::Call, nameToken->text);
  if (IS_TYPE(tokens, TokenType::RP))
    tokens.pop_front();
  auto child = primary(tokens);
  if (child != nullptr)
    node->addChild(child);
  while (!tokens.empty() && tokens.front()->type != TokenType::RP)
  {
    auto child = primary(tokens);
    if (child != nullptr)
      node->addChild(child);
  }
  if (tokens.empty())
    throw "call args syntax error";
  tokens.pop_front();
  return node;
};
/**
 * 一元前缀表达式 unaryExpression :: [!,~] primary
 */
ASTNode *unaryExpression(list<Token *> &tokens)
{
  LOG_FRONT("unaryExpression", tokens);
  ASSET_TS(tokens, TokenType::UNARYOP);
  auto node = new ASTNode(ASTNodeType::Unary, tokens.front()->text);
  tokens.pop_front();
  auto child = primary(tokens);
  if (child == nullptr)
    throw "mising primary";
  node->addChild(child);
  return node;
};
/**
 * 乘除法表达式
 * multiplicative :: primary ( '*'|'/' primary)*
 * @param tokens
 */
ASTNode *multiplicative(list<Token *> &tokens)
{
  LOG_FRONT("multiplicative", tokens);
  auto child1 = primary(tokens);
  if (child1 == nullptr)
    return child1;

  auto node = child1;
  while (!tokens.empty())
  {
    if (!IS_TYPE(tokens, TokenType::MULOP))
      break;
    auto opToken = tokens.front();
    tokens.pop_front();
    auto child2 = primary(tokens);
    if (child2 == nullptr)
      throw "invalid additive expression, expecting the right part.";
    node = new ASTNode(ASTNodeType::Binary, opToken->text);
    node->addChild(child1);
    node->addChild(child2);
    child1 = node;
  }
  return node;
}
/**
 * 加减法表达式：+-属于加
 * additiveExpression :: multiplicative ( '+'|'-' multiplicative)*
 * @param tokens
 */
ASTNode *additive(list<Token *> &tokens)
{
  LOG_FRONT("additive", tokens);
  auto child1 = multiplicative(tokens);
  if (child1 == nullptr)
    return child1;

  auto node = child1;
  while (!tokens.empty())
  {
    auto opToken = tokens.front();
    if (!IS_TYPE(tokens, TokenType::PLUS) && !IS_TYPE(tokens, TokenType::MINUS))
      break;
    tokens.pop_front();
    auto child2 = multiplicative(tokens);
    if (child2 == nullptr)
      throw "invalid additive expression, expecting the right part.";
    node = new ASTNode(ASTNodeType::Binary, opToken->text);
    node->addChild(child1);
    node->addChild(child2);
    child1 = node;
  }
  return node;
}
/**
 * 二元表达式 BinaryExpression :: [<<,>>,==,!=,<,<=,>,>=,additive]
 */
ASTNode *binaryExpression(list<Token *> &tokens)
{
  LOG_FRONT("binaryExpression", tokens);
  auto child = additive(tokens);
  if (tokens.empty() || child == nullptr || child->type == ASTNodeType::Binary)
    return child;

  auto type = tokens.front()->type;
  //说明 child 是普通的 primary 可与后面的结合成二元表达式
  switch (type)
  {
  case TokenType::SHOP:   //<< >>
  case TokenType::RELOP:  //< > <= >=
  case TokenType::EQOP:   //== !=
  case TokenType::OR:     // ||
  case TokenType::AND:    // &&
  case TokenType::BITOR:  // |
  case TokenType::BITXOR: // ^
  case TokenType::BITAND: // &
  {
    auto opToken = tokens.front();
    tokens.pop_front();
    auto child2 = additive(tokens);
    if (child2 == nullptr)
      throw "invalid additive expression, expecting the right part.";
    auto node = new ASTNode(ASTNodeType::Binary, opToken->text);
    node->addChild(child);
    node->addChild(child2);
    return node;
  }
  default:
    break;
  }
  return child;
};
ASTNode *expressionStatement(list<Token *> &tokens)
{
  LOG_FRONT("expressionStatement", tokens);
  vector<long> funcs = {
      (long)&callExpression,
      (long)&unaryExpression,
      (long)&binaryExpression};
  for (auto func : funcs)
  {
    auto node = ((Statement)func)(tokens);
    if (node != nullptr)
      return node;
  }
  return nullptr;
};
/**
 * 块级语句 blackStatement :: '{' (statement)* '}'
 */
ASTNode *blockStatement(list<Token *> &tokens)
{
  LOG_FRONT("blockStatement", tokens);
  ASSET_TS(tokens, TokenType::LC);
  auto node = new ASTNode(ASTNodeType::Block, "{}");
  tokens.pop_front();
  while (!tokens.empty())
  {
    if (IS_TYPE(tokens, TokenType::RC))
    {
      tokens.pop_front();
      return node;
    }
    auto child = statement(tokens);
    if (child != nullptr)
      node->addChild(child);
  }
  throw "block missing }";
};
/**
 * 表达式语句 returnStatement :: return expressionStatement?
 */
ASTNode *returnStatement(list<Token *> &tokens)
{
  LOG_FRONT("returnStatement", tokens);
  ASSET_TS(tokens, TokenType::RETURN);
  auto node = new ASTNode(ASTNodeType::Return, tokens.front()->text);
  tokens.pop_front();
  auto child = expressionStatement(tokens);
  if (child != nullptr)
    node->addChild(child);
  return node;
};
/**
 * 函数声明语句 functionDefinition :: function Identifier '(' Identifier (',' Identifier)* ')' blockStatement
 */
ASTNode *functionDefinition(list<Token *> &tokens)
{
  ASSET_TS(tokens, TokenType::FUNCTION);
  tokens.pop_front();
  if (!IS_TYPE(tokens, TokenType::NAME))
    throw "function must have name";
  auto nameToken = tokens.front();
  tokens.pop_front();
  if (!IS_TYPE(tokens, TokenType::LP))
    throw "function define error";
  auto node = new ASTNode(ASTNodeType::Function, nameToken->text);
  while (!tokens.empty())
  {
    if (IS_TYPE(tokens, TokenType::RP))
    {
      tokens.pop_front();
      auto child = blockStatement(tokens);
      if (child == nullptr)
        throw "function must have body";
      node->addChild(child);
      return node;
    }
    if (tokens.front()->type == TokenType::NAME)
      node->params.push_back(tokens.front()->text);
    tokens.pop_front();
  }
  return nullptr;
};
vector<long> getRootFuncs()
{
  vector<long> funcs = {
      (long)&varDeclare,
      (long)&assignmentStatement,
      (long)&expressionStatement,
      (long)&blockStatement,
      (long)&returnStatement,
      (long)&functionDefinition,
      (long)&callExpression};
  return funcs;
}